### PR TITLE
Fix DWARF composite layout in stack

### LINKED ISSUE #3845: test_vector_swap SIGTRAP
https://github.com/rizinorg/rizin/blob/4dd4dcb6c48bb269f7ff422cfb77e91bf447bc11/test/unit/test_vector.c#L654

```
>>> MALLOC_PERTURB_=224 /Users/aya/CLionProjects/rizin/build/test/unit/test_vector
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ✀  ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Backtrace 7 stack frames.
0   librz_util.0.7.dylib                0x000000010058fbbc rz_sys_backtrace + 44
1   librz_util.0.7.dylib                0x000000010051a9c4 rz_assert_log + 88
2   librz_util.0.7.dylib                0x00000001005a77d8 rz_vector_swap + 156
3   test_vector                         0x0000000100316930 test_vector_swap + 2816
4   test_vector                         0x00000001003032a0 all_tests + 1576
5   test_vector                         0x0000000100302c6c main + 28
6   dyld                                0x00000001a038bf28 start + 2236
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1/1 rizin:unit / vector FAIL            0.01s   killed by signal 5 SIGTRAP
```

### LINKED ISSUE #4005: DWARF composite layout in stack - problem parsing
If we have the following DWARF structure:
```
0x017dc4a0:     DW_TAG_variable
                  DW_AT_name	("this_status")
                  DW_AT_type	(0x017d51e3 "some_status_t")
                  DW_AT_location	(DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x1, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x1, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x1, DW_OP_piece 0x1, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x2, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x2, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x1, DW_OP_piece 0x1, DW_OP_lit0, DW_OP_stack_value, DW_OP_piece 0x2)
```
After parsing the variable it shows this as `afvl` and `afvll` output:
```
var some_status_t this_status @ COMPOSITE
var some_status_t this_status origin=DWARF @ composite: [(.0, 8): ,	(.0, 8): ,	(.0, 8): ,	(.0, 8): empty,	(.0, 16): ,	(.0, 16): ,	(.0, 8): ,	(.0, 8): empty,	(.0, 16): ]
```
Here the `some_status_t` is the structure type:
```
0x017d5127:   DW_TAG_structure_type
                DW_AT_name	("some_status_t")
                DW_AT_byte_size	(0x0c)
                DW_AT_sibling	(0x017d51e3)

0x017d514f:     DW_TAG_member
                  DW_AT_name	("memb0")
                  DW_AT_type	(0x017a53ea "uint8_t")
                  DW_AT_data_member_location	(0x00)

0x017d515c:     DW_TAG_member
                  DW_AT_name	("memb1")
                  DW_AT_type	(0x017a53ea "uint8_t")
                  DW_AT_data_member_location	(0x01)

0x017d5169:     DW_TAG_member
                  DW_AT_name	("memb2")
                  DW_AT_type	(0x017a53ea "uint8_t")
                  DW_AT_data_member_location	(0x02)

0x017d517f:     DW_TAG_member
                  DW_AT_name	("memb3")
                  DW_AT_type	(0x017a53f9 "uint16_t")
                  DW_AT_data_member_location	(0x04)

0x017d518c:     DW_TAG_member
                  DW_AT_name	("memb4")
                  DW_AT_type	(0x017a53f9 "uint16_t")
                  DW_AT_data_member_location	(0x06)

0x017d51a6:     DW_TAG_member
                  DW_AT_name	("memb5")
                  DW_AT_type	(0x017a53ea "uint8_t")
                  DW_AT_data_member_location	(0x08)

0x017d51c6:     DW_TAG_member
                  DW_AT_name	("memb6")
                  DW_AT_type	(0x017a53f9 "uint16_t")
                  DW_AT_data_member_location	(0x0a)
```
You can see, that it located completely at stack, so it should be reflected in the "COMPOSITE" string representation.

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

closes #4005
closes #3845

...

### CODE REVIEW

File: librz/util/vector.c [hunk 1]
```diff
@@ -85,23 +85,27 @@ RZ_API void rz_vector_free(RzVector *vec) {
  * \param src The vector to clone from.
  * \return true on success, false on failure.
  */
-RZ_API bool rz_vector_clone_into(
+RZ_API bool rz_vector_clone_intof(
```
Comment (XVilka) [inline librz/util/vector.c:89]: Please also keep the original functions as well, you removed them, also update Doxygen

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
Also:
```
cc-12 -Ilibrz/type/librz_type.so.0.7.0.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -I../librz/type/parser -Ilibrz/util/sdb/src -I../librz/util/sdb/src -I../subprojects/tree-sitter-c/src/tree_sitter -I../subprojects/tree-sitter-0.20.7/lib/src -I../subprojects/tree-sitter-0.20.7/lib/include -fdiagnostics-color=always -fsanitize=address,undefined -fno-omit-frame-pointer -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -O2 -g -Wimplicit-fallthrough=3 -DRZ_PLUGIN_INCORE=1 -D_GNU_SOURCE --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DASAN=1 -DRZ_ASSERT_STDOUT=1 -Wno-cpp -ftrivial-auto-var-init=pattern -funsigned-char -fPIC -MD -MQ librz/type/librz_type.so.0.7.0.p/base.c.o -MF librz/type/librz_type.so.0.7.0.p/base.c.o.d -o librz/type/librz_type.so.0.7.0.p/base.c.o -c ../librz/type/base.c
../librz/type/base.c: In function ‘rz_base_type_clone_into’:
../librz/type/base.c:163:85: error: passing argument 3 of ‘rz_vector_clone_intof’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  163 |                 rz_vector_clone_intof(&dst->enum_data.cases, &src->enum_data.cases, RzTypeEnumCase_cpy);
      |                                                                                     ^~~~~~~~~~~~~~~~~~
      |                                                                                     |
      |                                                                                     _Bool (*)(RzTypeEnumCase *, RzTypeEnumCase *) {aka _Bool (*)(struct rz_type_enum_case_t *, struct rz_type_enum_case_t *)}
In file included from ../librz/include/rz_th.h:14,
                 from ../librz/include/rz_util.h:13,
                 from ../librz/type/base.c:4:
../librz/include/rz_vector.h:74:28: note: expected ‘void (*)(void *, void *)’ but argument is of type ‘_Bool (*)(RzTypeEnumCase *, RzTypeEnumCase *)’ {aka ‘_Bool (*)(struct rz_type_enum_case_t *, struct rz_type_enum_case_t *)’}
   74 |         RZ_NULLABLE void (*item_cpy)(void *, void *));
      |                     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
../librz/type/base.c:166:93: error: passing argument 3 of ‘rz_vector_clone_intof’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  166 |                 rz_vector_clone_intof(&dst->struct_data.members, &src->struct_data.members, RzTypeStructMember_cpy);
      |                                                                                             ^~~~~~~~~~~~~~~~~~~~~~
      |                                                                                             |
      |                                                                                             _Bool (*)(RzTypeStructMember *, RzTypeStructMember *) {aka _Bool (*)(struct rz_type_struct_member_t *, struct rz_type_struct_member_t *)}
../librz/include/rz_vector.h:74:28: note: expected ‘void (*)(void *, void *)’ but argument is of type ‘_Bool (*)(RzTypeStructMember *, RzTypeStructMember *)’ {aka ‘_Bool (*)(struct rz_type_struct_member_t *, struct rz_type_struct_member_t *)’}
   74 |         RZ_NULLABLE void (*item_cpy)(void *, void *));
      |                     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
../librz/type/base.c:169:91: error: passing argument 3 of ‘rz_vector_clone_intof’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  169 |                 rz_vector_clone_intof(&dst->union_data.members, &src->union_data.members, RzTypeUnionMember_cpy);
      |                                                                                           ^~~~~~~~~~~~~~~~~~~~~
      |                                                                                           |
      |                                                                                           _Bool (*)(RzTypeUnionMember *, RzTypeUnionMember *) {aka _Bool (*)(struct rz_type_union_member_t *, struct rz_type_union_member_t *)}
../librz/include/rz_vector.h:74:28: note: expected ‘void (*)(void *, void *)’ but argument is of type ‘_Bool (*)(RzTypeUnionMember *, RzTypeUnionMember *)’ {aka ‘_Bool (*)(struct rz_type_union_member_t *, struct rz_type_union_member_t *)’}
   74 |         RZ_NULLABLE void (*item_cpy)(void *, void *));
      |                     ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
../librz/type/base.c: In function ‘RzTypeUnionMember_cpy’:
../librz/type/base.c:145:1: error: control reaches end of non-void function [-Werror=return-type]
  145 | }
      | ^
../librz/type/base.c: In function ‘RzTypeStructMember_cpy’:
../librz/type/base.c:132:1: error: control reaches end of non-void function [-Werror=return-type]
  132 | }
      | ^
../librz/type/base.c: In function ‘RzTypeEnumCase_cpy’:
../librz/type/base.c:138:1: error: control reaches end of non-void function [-Werror=return-type]
  138 | }
      | ^
cc1: all warnings being treated as errors
```

**Comment by XVilka** [comment]
@imbillow still shows:
```
var some_status_t this_status origin=DWARF @ composite: [(.0, 8): ,	(.0, 8): ,	(.0, 8): ,	(.0, 8): empty,	(.0, 16): ,	(.0, 16): ,	(.0, 8): ,	(.0, 8): empty,	(.0, 16): ]
```

File: librz/include/rz_vector.h [hunk 2]
```diff
@@ -68,11 +68,31 @@ RZ_API void rz_vector_fini(RzVector *vec);
...
-	RZ_NONNULL RZ_BORROW RZ_IN RzVector *src);
+	RZ_NONNULL RZ_BORROW RZ_IN RzVector *src,
+	RZ_NULLABLE void (*item_cpy)(void *, void *));
 
-RZ_API RzVector *rz_vector_clone(RzVector *vec);
+RZ_API RZ_OWN RzVector *rz_vector_clonef(
+	RZ_NONNULL RZ_BORROW RZ_IN RzVector *vec,
+	RZ_NULLABLE void (*item_cpy)(void *, void *));
+
+static inline bool rz_vector_clone_into(
```
Comment (b1llow) [inline librz/include/rz_vector.h]: @XVilka In fact, the original rz_vector_clone_into and rz_vector_clone were not removed, they were here

Comment (XVilka) [reply librz/include/rz_vector.h]: But the API mark was removed still

**Review by XVilka — CHANGES_REQUESTED** [review]
Fixed the problem but fails to compile on Windows:
```
"cl" "-Ilibrz\util\rz_util.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Isubprojects\zlib-1.3" "-I..\subprojects\zlib-1.3" "-I..\subprojects\xz-5.4.3\src\liblzma\api" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util.lib.p\vector.c.pdb" /Folibrz/util/rz_util.lib.p/vector.c.obj "/c" ../librz/util/vector.c
../librz/util/vector.c(109): error C2036: 'void *': unknown size
../librz/util/vector.c(109): error C2036: 'void *const ': unknown size
[441/2044] Compiling C object librz/util/rz_util.lib.p/utf32.c.obj
[442/2044] Compiling C object librz/util/rz_util.lib.p/utf8.c.obj
ninja: build stopped: subcommand failed.
```

**Pull request merged**